Add data plane REST API for Azure SignalR

vicancy opened the pull request:
Azure SignalR currently has its own [Management API](https://github.com/Azure/azure-signalr/blob/dev/docs/management-sdk-guide.md) to manage the clients connected to the service. Underneath it has both persistent connection mode and transient(REST API) mode.

We'd like to leverage azure-sdk to generate the data-plane REST API clients:
1. The current self-written REST API client used by the Management API can be replaced with the C# client azure-sdk generated
2. Customers using other languages, e.g. JavaScript, Python, can also leverage the generated client to talk to the service more conveniently

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

vicancy committed:
Add data plane REST API for Azure SignalR

Azure SignalR currently has its own [Management API](https://github.com/Azure/azure-signalr/blob/dev/docs/management-sdk-guide.md) to manage the clients connected to the service. Underneath it has both persistent connection mode and transient(REST API) mode.

We'd like to leverage azure-sdk to generate the data-plane REST API clients:
1. The current self-written REST API client used by the Management API can be replaced with the C# client azure-sdk generated
2. Customers using other languages, e.g. JavaScript, Python, can also leverage the generated client to talk to the service more conveniently

AutorestCI commented:
/azp run automation - sdk

azuresdkci commented:
Can one of the admins verify this patch?

vicancy committed:
Create readme.md

AutorestCI commented:
/azp run automation - sdk

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+              "description": "Ok"
+            },
+            "503": {
+              "description": "ServiceUnavailable"
```
We suggest to not document error, or to mark them with `x-ms-error-response` so the codegen knows this is an error. Otherwise SDK here will consider a 503 a success, which is unlikely what you want.
https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-error-response

vicancy replied:
Nice catch, I will remove all the 5xx error codes.

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+    "paths": {
+      "/api/v1/health": {
+        "get": {
+          "description": "Get service health status",
```
This is actually a summary

vicancy replied:
Are there any rules for distinguishing `summary` and `description`? Does it matter if it is `description`? Actually the swagger is auto-generated from `Swashbuckle Swagger` and the `description` is extracted from **triple-slash comments**'s `<summary>`  directly.

lmazuel replied:
It's not critical, since most of the SDK would be fine if they have either a summary or a description. If you auto-generates this file and there is no configuration, it's not worth the effort.

vicancy replied:
Updated to summary

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+      "/api/v1/health": {
+        "get": {
+          "description": "Get service health status",
+          "tags": [],
```
Please add OperationId

vicancy replied:
Sure thing, will add.

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+          },
+          "responses": {
+            "202": {
+              "description": "Accepted"
```
Does your 202 have a way to poll the result that is compatible with Microsoft OenAPI guidelines?

vicancy replied:
No, for sending messages `202` does not have corresponding polls, it **fire and forget**. Sending messages through REST API, including *send to connection, send to user/group and broadcast* follows **fire and forget** pattern. The service does not track if every subscriber has recieved the message and there is no plan to support that yet. (Customers might implement their own message protocol if they want to guarantee message delivery, for example using acks)

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+              "description": "Ok"
+            },
+            "404": {
+              "description": "NotFound"
```
See previous comment on x-ms-error-response

vicancy replied:
Hey @Imazuel, thank you for taking the time to look into the PR and provide valuable feedbacks!

For this `404` one, this API is to check existence, and we leverage `404` to indicate that the resource does not exist. So the semantics of `404` is different than 5xx errors. Is `x-ms-error-response` also applicable in this case?

lmazuel replied:
I saw a GET verb, so I assumed you wanted to return a User or nothing. I see you indeed didn't have a schema, so I guess it's fair.
If your API supports it, checks are usually models using HEAD, so save some body on the wire (just saying).

vicancy replied:
We add both `HEAD` and `GET`, for `GET`, the difference is the error response contains detailed error when `400 Bad Request`.  Shall we add `x-ms-error-response` to `4xx` responses?

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+              "description": "Ok"
+            },
+            "404": {
+              "description": "NotFound"
```
See previous comment on x-ms-error-response

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,710 @@
+      "schemas": {
+        "IActionResult": {
+          "type": "object",
+          "additionalProperties": false
```
Shouldn't be necessary?

vicancy replied:
Nice catch, will update.

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Update readme.md

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Change to use 2.0 swagger

vicancy committed:
Fix the format of operationId

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Fix autorest errors and warnings

vicancy committed:
Suppress warning that does not match the case

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Fix prettier

AutorestCI commented:
/azp run automation - sdk

vicancy commented:
Hi @lmazuel would you mind taking another round of review for the latest changes passing all the checks? Thanks for your time!

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+          "general"
+        ],
+        "summary": "Get service health status.",
+        "operationId": "HealthApiIndex_Head",
```
Operation name should not show the implementation detail of the HTTP layer. We could call it `HealthApiIndex_GetServiceHealthStatus` for instance to be more explicit. The API review board might have stronger naming opinion.

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+          "signalr.v1"
+        ],
+        "summary": "Broadcast a message to all clients connected to target hub.",
+        "operationId": "HubProxyControllerV1Broadcast_Post",
```
Same naming strategy applies here (don't call it POST). This applies to all operation that follow this HTTP naming pattern in the file.

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+          {
+            "in": "path",
+            "name": "hub",
+            "description": "Target hub name, which should start with alphabetic characters and only contain alpha-numeric characters or underscore.",
```
To check with review board if they want it: you can use the `pattern` node to explain in regexp your constraints.

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+            "in": "query",
+            "name": "excluded",
+            "description": "Excluded connection Ids",
+            "type": "array",
```
By default, this will be serialized as `excluded=id1,id2` (csv collectionFormat). If you have a different serialization format, please use this doc to explain the right value:
https://swagger.io/docs/specification/2-0/describing-parameters/ (section *Array and Multi-Value Parameters*)

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+          },
+          {
+            "in": "body",
+            "name": "body",
```
Same applies to parameter name, this should not be `body` but something like `payloadMessageParameter`

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+          "202": {
+            "description": "Success"
+          },
+          "400": {
```
If you declare it like that, this won't be considered an error, but a success. You can either remove totally that node and let the SDK handle it (there is default behavior that are probably fine), or use x-ms-error-response extension if you really care about defining it explicitly:
https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-error-response

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,875 @@
+    }
+  },
+  "definitions": {
+    "PayloadMessage": {
```
Is anything required, or all is optional?

vicancy replied:
All the payload messages are required

lmazuel replied:
I meant, are `target` and `arguments` required for a `PayloadMessage` structure?

vicancy replied:
nice catch! `target` should be `required`

vicancy committed:
Address comments

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

lmazuel commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,742 @@
+          "202": {
+            "description": "Success"
+          },
+          "400": {
```
Should 404 be added, so a delete on a non-existing user is not an error?

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+            "description": "Target hub name, which should start with alphabetic characters and only contain alpha-numeric characters or underscore.",
+            "required": true,
+            "type": "string",
+            "pattern": "^[A-Za-z][A-Za-z0-9_`,.[\\]]{0,127}$"
```
Should the period (`.`) be escaped?

vicancy replied:
no need to because it is inside `[]`. should remove one escape `\` used inside c# though.

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+        "arguments": {
+          "type": "array",
+          "items": {
+            "type": "object"
```
Is there no known structure to the items in the array? In other words, they can be arbitrary key/value pairs (but not strings etc.)?

vicancy replied:
Yes the items of `PayloadMessage` can be any object, there is no constraint on the object.

johanste replied:
So they could be strings? If so, `"type": "object"` should be removed since that implies that it has to be a mapping like object (key/value pairs)

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+            }
+          }
+        ],
+        "responses": {
```
Is there no response body in the case of failures? If there is, its schema should be included in the swagger. And even if there is no schema, I would include a `"default"` section explicitly calling out the fact that there is no response body.

vicancy replied:
Quoted from @lmazuel , as the comment for my previous PR iteration which contains explicity error responses with `text/plain` body descriptions. The comment makes sense to me because they are not that informative and the error code to the most extent explains why it fails. So that's the reason I removed the error nodes.
> •	You preferred to use “x-ms-error-response” for the 400 errors that returns text/plain errors. Is there a specific reason why you chose to use the extension, and not just removing the status code completely from the Swagger? I know I gave you the two options, but I was slightly encouraging the “don’t say anything” approach. I don’t feel most of them are informative from a documentation perspective (it’s just say 400 – Bad request, which is at the HTTP spec level), and that makes you add “text/plain” in the “produces” node, which I’d rather avoid (since the success path does not return “text/plain”). So my advice would be more explicit about removing all those, and let azure-core in all languages deal with 4xx automatically.

johanste replied:
Why are you returning text/plain response bodies for errors? The Microsoft REST API Guidelines have a recommended json format for error responses that allow clients to programmatically understand what the error was. A 400 response basically only says "you did something wrong", which makes it more or less useless from a client perspective for anything other than logging it.

vicancy replied:
Thanks for the insights! will follow https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#7102-error-condition-responses to add error response

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
A 200 response with no response body? If so, it should normally be a 204 status code - is there a reason that you are returning 200 rather than 204?

vicancy replied:
I will add the body description (`text/plain`) if that is a concern from your side.

johanste replied:
Does the service return text in the response body? If so, yes, that body description should be there. If it doesn't return anything in the response body, the status code should be 204 rather than 200 (and consequently, it wouldn't have a response body schema, since you are not returning anything)

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "Success"
```
No response body? If so, this should probably return a 204.

vicancy replied:
I will add a content type "plain/text"

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
Is there any way to monitor if the close actually succeeded? This applies to all methods that return 202s - they may eventually fail, and the client has no way (that I can see in the API) to determine if the operation succeeded or failed. Are they expected to call the "check for existence" after some random time until the connection is gone or does it not really matter if the close succeeds or fails (from the calling client's perspective)?

vicancy replied:
for now, no, it follows the "fire and forget" pattern as similar to SignalR's concept.

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+      }
+    },
+    "/api/v1/hubs/{hub}/groups/{group}/connections/{connectionId}": {
+      "put": {
```
Having a PUT with no request body is unusual (although not explicitly disallowed by the HTTP spec). What Content-Type and Content-Length does the service assume for this request?

vicancy replied:
Semantically I think `PUT` fits the scenario to add a connection to the group. The current service implementation does not check `Content-Length` and `Content-Type`.

vicancy replied:
Any concerns about current design @johanste ?

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+          "200": {
+            "description": "Success"
+          },
+          "404": {
```
What does the service return if the hub, connection and group (respectively) cannot be found?

vicancy replied:
404 returned when the connection is not found in the group. So for any of the cases, the assertion "connection is not in the group" is true, as a result, all these return 404.

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,605 @@
+    }
+  },
+  "securityDefinitions": {
+    "Bearer": {
```
More of a tooling question; this is technically not the correct representation - but unfortunately, I don't think that openapi 2.0 supports what you are trying to do. @lmazuel, will autorest and friends do the "right thing" with this?

vicancy replied:
Then shall I remove this `securityDefinitions` section or keeping it is preferred?

vicancy committed:
resolve comments

1. add detailed schema for 200
2. add default error section

AutorestCI commented:
/azp run automation - sdk

johanste commented on specification/signalr/data-plane/AzureSignalR/v1/azuresignalr.json:
```diff
@@ -0,0 +1,721 @@
+          }
+        }
+      },
+      "get": {
```
If the only thing you want to convey, I would recommend that you use the HTTP HEAD method rather than GET and leave the door open for (at some point in the future) return connection related metadata.

vicancy replied:
sounds super great solution

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

vicancy committed:
Update azuresignalr.json

AutorestCI commented:
/azp run automation - sdk

vicancy commented:
This PR is ready for review, all the comments from the review meeting should be resolved now. 

@johanste please take a look to see if there are any other concerns.

vicancy committed:
rename version

vicancy committed:
Fix build failure

vicancy committed:
Update azuresignalr.json

Update all DELETE actions to 200 response,
and update to apply to `:send` action for POST messages

vicancy committed:
Update azuresignalr.json
